### ISSUE TITLE
Calls to functions which may change the state of the floating-point environment may be optimized out

### ISSUE THREAD

**Issue by PavelMorozkin** [llvm/llvm-project#52891]
Sample code:
```
#include <fenv.h>
#include <math.h>

#if __STDC__ == 1 && __STDC_IEC_559__ == 1

#pragma STDC FENV_ACCESS ON

int main(void)
{
    sin(0.5);
    return fetestexcept(FE_INEXACT) ? 0 : 1;
}

#endif
```
Options: `-O1 -std=c11 -pedantic -Wall -Wextra -lm -ffp-model=strict`
Generated code:
```
main: # @main
        push rax
        mov edi, 32
        call fetestexcept
```
Here we see that `call sin` was incorrectly optimized out.

**Comment by pmor13** [comment]
@AaronBallman FYI: the bug is still present.

**Comment by AaronBallman** [comment]
There's definitely a bug here, but I'm not yet certain *where* the bug is. Clang does not define `__STDC_IEC_559__` so we don't claim any conformance to the Annex F rules. However, stdc-predef.h defines it whether the compiler wants or not: https://godbolt.org/z/oddWazvb7
```
<snip>
# 1 "/usr/include/stdc-predef.h" 1 3 4
# 19 "/usr/include/stdc-predef.h" 3 4
#define _STDC_PREDEF_H 1
# 41 "/usr/include/stdc-predef.h" 3 4
#define __STDC_IEC_559__ 1

#define __STDC_IEC_559_COMPLEX__ 1

#define __STDC_ISO_10646__ 201706L
# 440 "/usr/include/features.h" 2 3 4
<snip>
```
So it seems like both glibc and musl define this macro regardless:
https://sourceware.org/git/?p=glibc.git;a=blob;f=include/stdc-predef.h;h=d76933d4682871702c4b127938273b6b11c8addd;hb=HEAD#l36

https://git.musl-libc.org/cgit/musl/tree/include/stdc-predef.h

This is not the first time we've had a fight between the compiler, the standard library, and the standard predefined macro requirements (we hit something similar with `__STDC_ISO_10646__` where the compiler and the library need to agree in order to set feature test macro value properly). I don't think we've found a satisfactory solution to that issue yet.

**Comment by pmor13** [comment]
@AaronBallman 

1. About `stdc-predef.h`: have a look: https://github.com/llvm/llvm-project/issues/53132.
2. Defining both `__STDC__` and `__STDC_IEC_559__` to 1 “in advance” (i.e. w/o a prior knowledge whether the (3rd party) C library is conforming) may be confusing because a non-conforming C library makes the whole implementation non-conforming.

Extra: MSVC has this issue too: the C11 support is claimed, however there is `fatal error C1083: Cannot open include file: 'stdalign.h': No such file or directory`. Answer from Mike Danes:
> It may be worth pointing out that stdalign.h (and pretty much all other C headers) are provided by the Windows SDK, not by VC++. Older versions of the Windows SDK may not have it so you can end up in a situation where the compiler supports C11 but the SDK does not.

**Comment by AaronBallman** [comment]
Yup, C's feature testing macros are problematic for Clang because we allow the user to pick the C standard library they want to use and so there's no way for the frontend to provide the predefined macros without lying to users. It is possible for the C standard library to provide the predefined macros in these cases, but that requires very careful tests for compiler versions to decide what to expose and with what value.

Ultimately, I think what needs to happen is that we need two sets of feature testing macros in some cases -- one for the library portion and one for the language portion (similar to how WG21 already handles feature testing). Then the library can use the language's feature test macro to determine what interfaces to support and the library can specify their feature test macro value accordingly.

**Comment by andykaylor** [comment]
I just came across this issue while looking for something else. It's an interesting case, and I don't think it's a bug.

I think this whole business about `__STDC_IEC_559__` is a red herring, because the fact is that we intend to support IEC 559 exception behavior. I understand we have some problems with it, but we intend for it to be there and whenever we have something wrong, it should be considered a bug.

However, in this case the optimizer is trying to do something clever. The front end generates the call to sin(). If you use "-fno-math-errno" it will generate a call to the llvm.experimental.constrained.sin.f64 intrinsic, but by default it generates the normal library call, because the call might set errno. The place where the call is eliminated might not be what you'd expect. What's happening is that LibCallsShrinkWrapPass thinks it knows when sin will encounter an error and inserts a check for it.

Before LibCallsShrinkWrapPass:

```
define dso_local i32 @main(i32 noundef %argc, ptr nocapture noundef readnone %argv) local_unnamed_addr {
entry:
  %call = call double @sin(double noundef 5.000000e-01)
  %call1 = call i32 @fetestexcept(i32 noundef 32)
  %tobool.not = icmp eq i32 %call1, 0
  %cond = zext i1 %tobool.not to i32
  ret i32 %cond
}
```
After LibCallsShrinkWrapPass:

```
define dso_local i32 @main(i32 noundef %argc, ptr nocapture noundef readnone %argv) local_unnamed_addr {
entry:
  %0 = call i1 @llvm.experimental.constrained.fcmp.f64(double 5.000000e-01, double 0xFFF0000000000000, metadata !"oeq", metadata !"fpexcept.strict")
  %1 = call i1 @llvm.experimental.constrained.fcmp.f64(double 5.000000e-01, double 0x7FF0000000000000, metadata !"oeq", metadata !"fpexcept.strict")
  %2 = or i1 %1, %0
  br i1 %2, label %cdce.call, label %cdce.end

cdce.call:                                        ; preds = %entry
  %call = call double @sin(double noundef 5.000000e-01)
  br label %cdce.end

cdce.end:                                         ; preds = %entry, %cdce.call
  %call1 = call i32 @fetestexcept(i32 noundef 32)
  %tobool.not = icmp eq i32 %call1, 0
  %cond = zext i1 %tobool.not to i32
  ret i32 %cond
}
```
That is, LibCallsShrinkWrap inserted code to check for inputs as +/- inf. Since the argument is constant, those compares will be folded away, and that's where the call is eliminated.

The question then becomes whether we are required to preserve any and all floating point exceptions that would have been raised if that call had been executed, or only the ones that are documented as expected. The C standard says that the sin() call will raise FE_INVALID if an infinity is passed as an argument. It doesn't say that the function will raise FE_INEXACT. I believe that is explicitly unspecified.

I'm not sure there's a bug here.

**Comment by jcranmer-intel** [comment]
> The question then becomes whether we are required to preserve any and all floating point exceptions that would have been raised if that call had been executed, or only the ones that are documented as expected. The C standard says that the sin() call will raise FE_INVALID if an infinity is passed as an argument. It doesn't say that the function will raise FE_INEXACT. I believe that is explicitly unspecified.

F.10p10:
>  Whether or when library functions not listed in the "Operation binding" table in F.3 raise the "inexact" floating-point exception is unspecified, unless stated otherwise.

Indeed, the raising of inexact for libm functions is explicitly unspecified.

OTOH, as a QoI issue, it does feel a little extreme to say "C standard says its unspecified, so we'll optimize all checks away from you", even if the user is (presumably) going to be careful to audit their library to see if it's reporting `FE_INEXACT` correctly.

**Comment by andykaylor** [comment]
> OTOH, as a QoI issue, it does feel a little extreme to say "C standard says its unspecified, so we'll optimize all checks away from you", even if the user is (presumably) going to be careful to audit their library to see if it's reporting `FE_INEXACT` correctly.

Yes, I can't disagree with this. Probably the shrink wrap code should be checking for strictfp.
